File: go-src/pkg/letterboxed/letterboxed.go
package letterboxed

import (
	"letter-boxed-solver/pkg/sets"
	"runtime"
	"strings"
	"sync"
)

// PrefixDict maps the first character of a word to all words that start with it.
type PrefixDict map[rune]sets.Set[string]

type SolveTask struct {
	previousWords []string
}

// LetterBoxed holds the dictionary and prefix dictionary for fast lookups.
type LetterBoxed struct {
	dict       sets.Set[string]
	prefixDict PrefixDict
}

// NewLetterBoxed constructs a [LetterBoxed] from a list of dictionary words and sides of the square.
func NewLetterBoxed(dict []string, sides []string) *LetterBoxed {
	allowedWords := getAllowedWords(dict, sides)
	prefixes := buildPrefixDict(allowedWords.ToSlice())

	return &LetterBoxed{
		dict:       allowedWords,
		prefixDict: prefixes,
	}
}

// Solutions finds all sequences of words that adhere to the rules of the game.
// It returns a read-only channel and launches a worker pool to send results.
func (s *LetterBoxed) Solutions(maxWords int) <-chan []string {
	out := make(chan []string)
	tasks := make(chan SolveTask, 100_000_000)
	var wg sync.WaitGroup

	// Start as many workers as there are CPU cores
	for i := 0; i < runtime.NumCPU(); i++ {
		go func() {
			for t := range tasks {
				s.workSubSolutions(t.previousWords, maxWords, out, tasks, &wg)
			}
		}()
	}

	// Seed initial task
	wg.Add(1)
	tasks <- SolveTask{previousWords: nil}

	go func() {
		wg.Wait()
		close(tasks)
		close(out)
	}()

	return out
}

func (s *LetterBoxed) workSubSolutions(
	previousWords []string,
	maxWords int,
	out chan<- []string,
	tasks chan<- SolveTask,
	wg *sync.WaitGroup,
) {
	var wordSet sets.Set[string]

	if len(previousWords) == 0 {
		wordSet = s.dict
	} else {
		lastWord := previousWords[len(previousWords)-1]
		// The last letter of the last word is the starting letter for the current word
		var startingLetter rune
		for _, ch := range lastWord {
			startingLetter = ch
		}

		// Words that were already used in the previous words won't help
		wordSet = s.prefixDict[startingLetter].Diff(sets.New(previousWords...))
	}
	if len(wordSet) == 0 {
		return
	}

	for word := range wordSet {
		newWords := append(previousWords, word)

		// If there are no more unused letters, it means we've found a solution
		if s.countUnusedLetters(newWords) == 0 {
			out <- newWords
		} else if len(newWords) < maxWords {
			wg.Add(1)
			tasks <- SolveTask{previousWords: newWords}
		}
	}

	wg.Done()
}

// getAllowedWords filters dictionary words that can be formed with side letters and game rules.
func getAllowedWords(dict []string, sides []string) sets.Set[string] {
	allowedLetters := sets.New([]rune(strings.Join(sides, ""))...)
	disallowedPairs := sets.New[string]()

	for _, side := range sides {
		for _, a := range side {
			for _, b := range side {
				disallowedPairs.Add(string([]rune{a, b}))
			}
		}
	}

	allowedWords := sets.New[string]()
	for _, word := range dict {
		if len(word) < 3 {
			continue
		}
		if !sets.New([]rune(word)...).IsSubsetOf(allowedLetters) {
			continue
		}

		valid := true
		for i := 0; i < len(word)-1; i++ {
			if disallowedPairs.Contains(word[i : i+2]) {
				valid = false
				break
			}
		}
		if valid {
			allowedWords.Add(word)
		}
	}
	return allowedWords
}

// countUnusedLetters counts the number of letters that are not used in the given words.
func (s *LetterBoxed) countUnusedLetters(words []string) int {
	unusedLetters := sets.New[rune]()

	for prefix := range s.prefixDict {
		unusedLetters.Add(prefix)
	}

	for _, word := range words {
		for _, ch := range word {
			unusedLetters.Remove(ch)
		}
	}

	return len(unusedLetters)
}

// buildPrefixDict constructs a map of starting letters to words.
func buildPrefixDict(words []string) PrefixDict {
	dict := make(PrefixDict)
	for _, word := range words {
		if len(word) == 0 {
			continue
		}

		// Get the first letter of the word as a rune
		var ch rune
		for _, c := range word {
			ch = c
			break
		}

		if _, ok := dict[ch]; !ok {
			dict[ch] = sets.New[string]()
		}
		dict[ch].Add(word)
	}
	return dict
}

File: py-src/letterboxed.py
from itertools import pairwise, product, chain
from collections.abc import Collection
from typing import Generator


def to_set[T](col: Collection[T]) -> set[T]:
    return col if isinstance(col, set) else set(col)


class LetterBoxed:
    def __init__(self, dictionary: Collection[str], sides: Collection[str]):
        self.dictionary = self.get_allowed_words(dictionary, sides)
        self.prefix_dict = self.get_prefix_dict(self.dictionary)

    def solutions(self, max_words: int) -> Generator[list[str], None, None]:
        yield from self.sub_solutions([], max_words)

    def sub_solutions(self, previous_words: list[str], max_words: int) -> Generator[list[str], None, None]:
        wordset: set[str]

        if previous_words:
            starting_letter = previous_words[-1][-1]
            wordset = self.prefix_dict[starting_letter]
        else:
            wordset = self.dictionary

        # Remove words that have already been used since they won't be helpful
        wordset = wordset - set(previous_words)
        unused_letters = self.allowed_letters() - set(chain.from_iterable(previous_words))

        # Consider all possible words, prioritized by the number of new letters they use
        for word in self.sort_words_by_letter_usage(wordset, unused_letters):
            new_unused_letters = unused_letters - to_set(word)

            # If there are no letters left to use, it means we've found a solution
            if len(new_unused_letters) == 0:
                yield [*previous_words, word]
                continue

            if len(previous_words) + 1 < max_words:
                yield from self.sub_solutions([*previous_words, word], max_words)

    @staticmethod
    def get_allowed_words(dictionary: Collection[str], sides: Collection[str]) -> set[str]:
        """
        Get all words from the dictionary that can be formed with the letters in `sides`.
        Following the rules of the game, two letters from the same side are not allowed
        to be used consecutively, and the minimum number of letters in a word is 2.
        """
        allowed_letters = set("".join(sides))
        disallowed_pairs = set(chain.from_iterable(product(s, repeat=2) for s in sides))
        return set(
            w
            for w in dictionary
            # Minimum word length is 3
            if len(w) >= 3
            # The word must only use letters from the allowed letters
            and set(w) <= allowed_letters
            # The word must not contain two consecutive letters from the same side
            and all(p not in disallowed_pairs for p in pairwise(w))
        )

    @staticmethod
    def get_letter_usage(word: str, counted_letters: Collection[str]) -> int:
        """
        Count the number of distinct letters from `counted_letters` present in `word`.
        """
        return len(set(word) & to_set(counted_letters))

    @staticmethod
    def get_prefix_dict(words: Collection[str]) -> dict[str, set[str]]:
        """
        Create a dictionary mapping starting letters to words that start with that letter.
        """
        prefix_dict: dict[str, set[str]] = {}

        for word in words:
            prefix = word[0]

            if prefix not in prefix_dict:
                prefix_dict[prefix] = set()

            prefix_dict[prefix].add(word)

        return prefix_dict

    def sort_words_by_letter_usage(self, words: Collection[str], counted_letters: Collection[str]) -> list[str]:
        """
        Sort words by the number of distinct letters from `counted_letters` present in each word.
        """
        counted_letters = to_set(counted_letters)
        letter_usages = {w: self.get_letter_usage(w, counted_letters) for w in words}
        return sorted(letter_usages, key=lambda w: letter_usages[w], reverse=True)

    def allowed_letters(self) -> set[str]:
        """
        Get the set of letters that can be used in the game.
        """
        return set(self.prefix_dict.keys())
